File: frontend/src/api/me.ts
import { api } from './client'

export type ThemePref = 'light' | 'dark' | 'auto'

export interface SocialLink {
  label: string
  url: string
}

export interface Me {
  id: number
  full_name: string
  username: string | null
  phone: string | null
  language: string
  theme: ThemePref
  onboarded: boolean
  avatar_url: string | null
  social_links: SocialLink[]
  active_projects: number
  completed_projects: number
}

export interface MePatch {
  full_name?: string
  phone?: string
  theme?: ThemePref
  language?: string
  onboarded?: boolean
  social_links?: SocialLink[]
}

export const getMe = () => api<Me>('/api/me')

export const patchMe = (body: MePatch) =>
  api<Me>('/api/me', { method: 'PATCH', body: JSON.stringify(body) })

export function uploadAvatar(file: File): Promise<Me> {
  const fd = new FormData()
  fd.append('file', file)
  return api<Me>('/api/me/avatar', { method: 'POST', body: fd })
}

export const deleteAvatar = () =>
  api<Me>('/api/me/avatar', { method: 'DELETE' })

File: backend/app/schemas/me.py
"""Joriy foydalanuvchi sxemalari."""

from typing import Literal

from pydantic import BaseModel, Field, field_validator


class SocialLink(BaseModel):
    label: str = Field(min_length=1, max_length=40)
    url: str = Field(min_length=1, max_length=300)

    @field_validator("url")
    @classmethod
    def _https_only(cls, v: str) -> str:
        if not v.startswith("https://"):
            raise ValueError("Havola https:// bilan boshlanishi kerak")
        return v


class MeRead(BaseModel):
    id: int
    full_name: str
    username: str | None
    phone: str | None
    language: str
    theme: str
    onboarded: bool
    # bo'sh bo'lsa frontend Telegram rasmini ishlatadi
    avatar_url: str | None
    social_links: list[SocialLink]
    active_projects: int
    completed_projects: int


class MeUpdate(BaseModel):
    full_name: str | None = Field(default=None, min_length=1, max_length=200)
    phone: str | None = Field(default=None, max_length=32)
    theme: Literal["light", "dark", "auto"] | None = None
    language: str | None = Field(default=None, min_length=2, max_length=8)
    onboarded: bool | None = None
    social_links: list[SocialLink] | None = Field(default=None, max_length=5)
